src/Http/Routing/src/Matching/DfaMatcherBuilder.cs

```diff
@@ -288,7 +331,57 @@ public DfaNode BuildDfaTree(bool includeLabel = false)
 
                             if (parent.Literals != null)
                             {
-                                nextParents.AddRange(parent.Literals.Values);
+                                // For a complex segment like this, we can evaluate the literals and avoid adding extra nodes to
+                                // the tree on cases where the literal won't ever be able to match the complex parameter.
+                                // For example, if we have a complex parameter {a}-{b}.{c?} and a literal "Hello" we can guarantee
+                                // that it will never be a match.
+                                foreach (var literal in parent.Literals.Keys)
+                                {
+                                    var routeValues = new RouteValueDictionary();
```

Comment:
How much do we care about allocating this RVD every time vs reusing it?

Reply:
Reuse it if possible.

Reply:
done